### PR TITLE
Feature/1178 user authentication

### PR DESCRIPTION
**Pull request by amyd11**
*What does this PR do?*
Sets up authentication on the front end to work with authentication setup on the backend

*Jira ticket number*
[ISPGBSS-1178](https://jiraent.cms.gov/browse/ISPGBSS-1178)

### Changes

- Set up `sessionStorage` to house authentication info (token and username). Logging in will save these to session storage and logging out will clear them. _Note: Session storage was chosen over context so that this data can persist across a browser refresh._
- Updated `RequestService` to add `Authorization` field to header when making any api calls- _except a login request!_
- Created a Private wrapper component that is used to set up AppRoutes. When this is used, the route is locked down unless the user is authenticated
- Refactored config file to clean up code and added an additional url
    - This means all other files that import config needed an updated import statement & config reference, so yes, it’s a lot of files!
- Cleaned up mocked header code and replaced with real time user data
- Username in header is a button that, when clicked, will initiate logout
- Added a login page which the user will get redirected to if they try to access any page without being logged in
- Minor styling changes, but updated h1 globally to be the font size matching all h1’s used in designs _(verified that this wouldn’t cause any regression with already implemented h1’s)_
- Added tests for Header and Login

### Testing
**Test this in conjunction with the backend authentication changes https://github.com/CMSgov/blueprint_api/pull/66**

Ensure the following:
- You cannot access the site when you are not logged in
- Visiting the /login page and logging in gives you access to the site and get data (such as on the /projects page)
- When logged in, the username displays as a link in the header
- When clicking the username in the header, you are logged out

### Dev Notes
- We will want to handle token timeout, but that will come in another PR to compartmentalize reviewing and merging some functionality

### CODE REVIEW

File: src/organisms/Header.jsx [hunk 1]
```diff
@@ -20,11 +27,11 @@ const Header = () => {
         <Link to={MAIN_ROUTES.COMPONENT_LIBRARY}>Component Library</Link>
         <div className="site-header-right">
           <Link to={MAIN_ROUTES.HELP}>Help</Link>
-          <span className="usa-media-block__img circle-4 bg-base display-flex flex-row flex-align-center flex-justify-center text-white">
-            {count}
-          </span>
-          {/* @TODO: change user to a link when it exists */}
-          <a href="/user">{userName}</a>
+          {username && (
```
Comment (rmoore-theta) [inline src/organisms/Header.jsx:30]: Are we not going to have any way for users to edit their information/update password? It seems a little odd that clicking on your username will log you out :/

File: src/services/RequestService.js [hunk 2]
```diff
@@ -1,32 +1,48 @@
...
       })
       .catch((err) => failureCallback && failureCallback(err));
   },
   post: async (url, body, callback, failureCallback) => {
+    let postConfig = authConfig;
+
+    // if post request is accessing login (auth) endpoint
+    // use generic config headers without Authorization info (since there isn't any yet)
+    const isLoginPost = url === configUrl.backendUrlAuth;
+    if (isLoginPost) postConfig = config;
```
Comment (rmoore-theta) [inline src/services/RequestService.js]: Could you please wrap this block with {}

**Review by rmoore-theta — COMMENTED** [review]
It would have been nice to break out the config refactor as a separate PR.

File: src/organisms/Header.jsx (hunk 1 above)
Comment (jauschalley) [reply src/organisms/Header.jsx:30]: This will be controlled by IDM.  We will be getting this info from CMS central systems and will not be altering it.

Comment (jauschalley) [reply src/organisms/Header.jsx:30]: The logging out thought is a good point and we can bring that up with HCD.

Comment (rmoore-theta) [reply src/organisms/Header.jsx:30]: I know at some point we plan to get this information from IDM, but currently we are getting it from our own system.

File: src/organisms/Header.test.js [hunk 3]
```diff
@@ -1,13 +1,100 @@
-import { render, screen } from "@testing-library/react";
+import { fireEvent, render, screen } from "@testing-library/react";
 import { MemoryRouter } from "react-router-dom";
 import Header from "./Header";
```
Comment (jauschalley) [inline src/organisms/Header.test.js:4]: Praise: I especially appreciate your tests here.

**Review by jauschalley — DISMISSED** [review]
Praise: I know this was a long process with lots of detours and I appreciate how much effort you put into figuring out a good way to test this.

File: src/organisms/Header.test.js [hunk 4]
```diff
@@ -1,13 +1,100 @@
...
 
+const sessionStorageMock = (() => {
+  let store = {};
```
Comment (jauschalley) [inline src/organisms/Header.test.js:6]: Praise: I especially appreciate your tests here.

File: src/organisms/Header.jsx (hunk 1 above)
Comment (jauschalley) [reply src/organisms/Header.jsx:30]: We will not being going to production with this login.  This is a temporary measure for piloting.  We will not be editing information that is officially held by the CMS identity broker.

Comment (amyd11) [reply src/organisms/Header.jsx:30]: For the interim, HCD did create designs for how to log out, which involves clicking on the username and then seeing a dropdown menu which you can then click to log out. This work along with any styling or extra features to the sign in page can be done in an additional story if we want to develop this sign in/sign out feature. That story will be able to be prioritized or deprioritized depending on IDM work.

File: src/services/RequestService.js (hunk 2 above)
Comment (amyd11) [reply src/services/RequestService.js]: This is a one line if statement that just doesn't need the brackets. Do you want to add it for preference/readability?

**Comment by amyd11** [comment]
> It would have been nice to break out the config refactor as a separate PR.

Agreed because it touches a lot of files, but it was necessary to modify the config so I could add an additional backend endpoint for this story.

File: src/services/RequestService.js (hunk 2 above)
Comment (rmoore-theta) [reply src/services/RequestService.js]: This is usually a best practice for readability.

Comment (amyd11) [reply src/services/RequestService.js]: Updated

**Review by rmoore-theta — DISMISSED** [review]

**Review by jauschalley — APPROVED** [review]

**Review by jwhite3 — APPROVED** [review]
🚀

**Pull request merged**